File: resources/final.py
"""
Finalizing script
  1. Calculate the percentage error
  2. Sorting the resultant array(dictionary)
  3. Print the final value
  4. (optional)return the final value
"""
class final:
    def __init__(self,param):
        self.param = param
        self.combine()

    def welcome(self):
        print("Finalizing results -")
        print()

    def sort(self,key):
        #input parameter(resultant values)
        copy = self.param

        for j in range(0,len(copy)-1):
            for i in range(0,len(copy)-1-j):
                if copy[i+1][key] > copy[i][key]:
                    temp = copy[i]
                    copy[i] = copy[i+1]
                    copy[i+1] = temp

    def err_ratio(self):
        #            Theoretical - Simulation(model)
        # % error = --------------------------------- * 100%
        #                    Theoretical
        a = self.param
        for i in range(0,len(a)):
            theory = a[i]["theory"]
            model = a[i]["model"]
            err = (theory-model)/theory
            a[i]["error"]=err
            err_per = err*100
            print("Calculation method:",a[i]["method"])
            print("    Theoretical value:",theory)
            print("    Simulation value :",model)
            print("    Percentage error :",str(round(err_per,2))+"%")
            print()

    def combine(self):
        # calculate % error
        self.err_ratio()
        
        # sorting the result dictionary; self.param[0]["model"] is 
        # the resultant value
        self.sort("model")
        code = self.param[0]["method"]
        if code == "scp":
            label = "Simple cubic packing"
        elif code == "bcp":
            label = "Body-centered cubic packing"
        elif code == "hcp":
            label = "Hexagonal cubic packing"
        elif code == "fcp":
            label = "Face-centered cubic packing"
        print("By simulation calculation,",label,"structure results in",
            "maximum capacity of the container.")
        print("The maximum capacity of the container is",
            self.param[0]["model"],"sphere(s).")
        print()
        return self.param[0]["model"]

File: docs/resources/js/final.js
/**
Finalizing script
  1. Calculate the percentage error
  2. Sorting the resultant array(dictionary)
  3. console.log the Final value
  4. (optional)return the Final value
*/

var Final = function(param){
    this.param = param
    this.combine()
}

Final.prototype.welcome = function(){
    console.log("Finalizing results -")
    console.log()
}

Final.prototype.sort = function(key){
    //input parameter(resultant values)
    var copy = this.param
    for (var j=0;j<copy.length-1;j++){
        for (var i=0;i<copy.length-1-j;i++){
            if (copy[i+1][key] > copy[i][key]){
                var temp = copy[i]
                copy[i] = copy[i+1]
                copy[i+1] = temp
            }
        }
    }
}

Final.prototype.err_ratio = function(){
    //            Theoretical - Simulation(model)
    // % error = --------------------------------- * 100%
    //                    Theoretical
    var a = this.param
    for (var i=0;i<a.length;i++){
        var theory = a[i]["theory"]
        var model = a[i]["model"]
        var err = (theory - model)/ theory
        a[i]["error"]=err
        var err_per = (err * 100).toFixed(2)
        console.log("Calculation method: "+a[i]["method"])
        console.log("    Theoretical value: "+theory)
        console.log("    Simulation value : "+model)
        console.log("    Percentage error : "+err_per+"%")
        console.log()
    }
}

Final.prototype.combine = function(){
    // calculate % error
    this.err_ratio()
    
    // sorting the result dictionary; this.param[0]["model"] is 
    // the resultant value
    this.sort("model")
    var code = this.param[0]["method"], label
    switch (code){
        case "scp":
        label = "Simple cubic packing"
        break
        case "bcp":
        label = "Body-centered cubic packing"
        break
        case "hcp":
        label = "Hexagonal cubic packing"
        break
        case "fcp":
        label = "Face-centered cubic packing"
    }
    console.log("By simulation calculation, "+label+" structure results in "+
        "maximum capacity of the container.")
    console.log("The maximum capacity of the container is "+
        this.param[0]["model"]+" sphere(s).")
    console.log()
    return this.param[0]["model"]
}

Final.prototype.getResult = function(){
    return this.param
}